### ISSUE TITLE
Compare Optimizers Performance on Neural Network and Linear Regression Models

### ISSUE THREAD

**Issue by SAUE49** [SAUE49/Clinic-RH#1]
## Objective
Implement and compare different optimizers for both Neural Network and Linear Regression models using California Housing and Boston Housing datasets.

### Evaluation Criteria
1. **Convergence Speed**
   - Track the number of epochs needed to reach convergence
   - Monitor training time
   - Plot loss curves for each optimizer

2. **Generalization Performance**
   - Measure validation metrics (MSE, MAE, R²)
   - Analyze overfitting/underfitting behavior
   - Compare test set performance

3. **Hyperparameter Sensitivity**
   - Test different learning rates
   - Analyze impact of batch size
   - Document stability across different random seeds

#### Datasets
1. **California Housing Dataset**
   ```python
   from sklearn.datasets import fetch_california_housing
   ```

2. **Boston Housing Dataset**
   ```python
   from sklearn.datasets import load_boston
   # Note: Consider using an alternative source as this dataset is deprecated in newer scikit-learn versions
   ```

#### Models to Implement
1. **Linear Regression**
   - Standard implementation
   - With different regularization options (L1, L2)

2. **Neural Network**
   - Simple architecture (2-3 layers)
   - Consistent architecture across optimizer tests

#### Optimizers to Test
1. **First-Order Methods**
   - SGD (Stochastic Gradient Descent)
   - SGD with Momentum
   - Adam
   - RMSprop
   - AdaGrad

2. **Second-Order Methods (for Linear Regression)**
   - L-BFGS

### Experimental Setup
1. **Data Preprocessing**
   - Standard scaling
   - Train/validation/test split (60/20/20)
   - Handle missing values if any

2. **Hyperparameter Ranges**
   - Learning rates: [0.1, 0.01, 0.001]
   - Batch sizes: [32, 64, 128]
   - Number of epochs: Start with 100, adjust based on convergence

3. **Metrics Collection**
   - Training loss
   - Validation loss
   - Training time
   - Final test performance

### Expected Deliverables
1. Jupyter notebook with implementation
2. Plots comparing:
   - Loss curves for each optimizer
   - Convergence speed comparison
   - Validation performance
   - Hyperparameter sensitivity analysis
3. Summary table of best configurations
4. Analysis report with recommendations

### Technical Requirements
- Python 3.8+
- PyTorch or TensorFlow
- scikit-learn
- pandas
- matplotlib/seaborn for visualization

### Additional Notes
- Ensure reproducibility by setting random seeds
- Document all hyperparameters used
- Save model checkpoints for best performing configurations
- Include error handling and logging
